diff -Naur acse_2.0.2/acse/parser.h acse_2.0.2_patched/acse/parser.h
--- acse_2.0.2/acse/parser.h	2024-12-11 22:02:22
+++ acse_2.0.2_patched/acse/parser.h	2024-12-11 22:02:55
@@ -26,6 +26,11 @@
   t_label *lExit; ///< Label to the first instruction after the loop.
 } t_whileStmt;
 
+typedef struct {
+  t_regID rLastValue; ///< Register containing the previous value of the var.
+  t_label *lLoop;     ///< Label to the beginning of the loop.
+} t_convergeStmt;
+
 /**
  * @}
  */
diff -Naur acse_2.0.2/acse/parser.y acse_2.0.2_patched/acse/parser.y
--- acse_2.0.2/acse/parser.y	2024-12-11 22:02:22
+++ acse_2.0.2_patched/acse/parser.y	2024-12-11 22:03:05
@@ -50,6 +50,7 @@
   t_label *label;
   t_ifStmt ifStmt;
   t_whileStmt whileStmt;
+  t_convergeStmt convergeStmt;
 }
 
 /*
@@ -77,6 +78,7 @@
 %token <label> DO
 %token <string> IDENTIFIER
 %token <integer> NUMBER
+%token <convergeStmt> CONVERGE
 
 /*
  * Non-terminal symbol semantic value type declarations
@@ -182,6 +184,7 @@
   | return_statement SEMI
   | read_statement SEMI
   | write_statement SEMI
+  | converge_statement
   | SEMI
 ;
 
@@ -303,6 +306,29 @@
     t_regID rTmp = getNewRegister(program);
     genLI(program, rTmp, '\n');
     genPrintCharSyscall(program, rTmp);
+  }
+;
+
+converge_statement
+  : CONVERGE var_id
+  {
+    // Reserve a register that will buffer the old value of the variable
+    $1.rLastValue = getNewRegister(program);
+    // Generate a label that points to the body of the loop
+    $1.lLoop = createLabel(program);
+    assignLabel(program, $1.lLoop);
+
+    // Generate code that, just before each execution of the loop body,
+    // saves the value of the variable in the register we reserved earlier
+    t_regID rVar = genLoadVariable(program, $2);
+    genADD(program, $1.rLastValue, REG_0, rVar);
+  } 
+  code_block
+  {
+    // Generate a branch that continues the loop if the variable's current
+    // value is different than its previous one.
+    t_regID rVar = genLoadVariable(program, $2);
+    genBNE(program, rVar, $1.rLastValue, $1.lLoop);
   }
 ;
 
diff -Naur acse_2.0.2/acse/scanner.l acse_2.0.2_patched/acse/scanner.l
--- acse_2.0.2/acse/scanner.l	2024-12-11 22:02:22
+++ acse_2.0.2_patched/acse/scanner.l	2024-12-11 22:02:22
@@ -81,6 +81,7 @@
 "return"                  { return RETURN; }
 "read"                    { return READ; }
 "write"                   { return WRITE; }
+"converge"                { return CONVERGE; }
 
 {ID}                      {
                             yylval.string = strdup(yytext);
diff -Naur acse_2.0.2/tests/converge/converge.src acse_2.0.2_patched/tests/converge/converge.src
--- acse_2.0.2/tests/converge/converge.src	1970-01-01 01:00:00
+++ acse_2.0.2_patched/tests/converge/converge.src	2024-12-11 22:02:22
@@ -0,0 +1,11 @@
+int a, b;
+a = 31;
+
+converge a {
+  a = a / 3;
+  write(a);
+}
+
+converge a {
+  write(a);
+}
