diff -Naur acse_2.0.2/acse/parser.y acse_2.0.2_patched/acse/parser.y
--- acse_2.0.2/acse/parser.y	2024-12-11 21:55:35
+++ acse_2.0.2_patched/acse/parser.y	2024-12-11 21:56:52
@@ -77,6 +77,7 @@
 %token <label> DO
 %token <string> IDENTIFIER
 %token <integer> NUMBER
+%token <ifStmt> IIF
 
 /*
  * Non-terminal symbol semantic value type declarations
@@ -177,6 +178,7 @@
 statement
   : assign_statement SEMI
   | if_statement
+  | iif_statement
   | while_statement
   | do_while_statement SEMI
   | return_statement SEMI
@@ -207,6 +209,32 @@
     // Generate a jump to the else part if the expression is equal to zero.
     $1.lElse = createLabel(program);
     genBEQ(program, $3, REG_0, $1.lElse);
+  }
+  code_block
+  {
+    // After the `then' part, generate a jump to the end of the statement.
+    $1.lExit = createLabel(program);
+    genJ(program, $1.lExit);
+    // Assign the label which points to the first instruction of the else part.
+    assignLabel(program, $1.lElse);
+  }
+  else_part
+  {
+    // Assign the label to the end of the statement.
+    assignLabel(program, $1.lExit);
+  }
+;
+
+/* An iif statements first assign than computes the expression, then jumps to
+ * the `else' part if the expression is equal to zero.
+ * Otherwise the `then' part is executed.
+ * After the `then' part the `else' part needs to be jumped over. */
+iif_statement
+  : IIF LPAR assign_statement SEMI exp RPAR
+  {
+    // Generate a jump to the else part if the expression is equal to zero.
+    $1.lElse = createLabel(program);
+    genBEQ(program, $5, REG_0, $1.lElse);
   }
   code_block
   {
diff -Naur acse_2.0.2/acse/scanner.l acse_2.0.2_patched/acse/scanner.l
--- acse_2.0.2/acse/scanner.l	2024-12-11 21:55:35
+++ acse_2.0.2_patched/acse/scanner.l	2024-12-11 21:55:35
@@ -81,6 +81,7 @@
 "return"                  { return RETURN; }
 "read"                    { return READ; }
 "write"                   { return WRITE; }
+"iif"                     { return IIF; }
 
 {ID}                      {
                             yylval.string = strdup(yytext);
diff -Naur acse_2.0.2/tests/iif/iif.src acse_2.0.2_patched/tests/iif/iif.src
--- acse_2.0.2/tests/iif/iif.src	1970-01-01 01:00:00
+++ acse_2.0.2_patched/tests/iif/iif.src	2024-12-11 21:55:35
@@ -0,0 +1,8 @@
+int i,j;
+i=5;
+j=20;
+iif (i=i+2; (j*5)>i) {
+  j = i;
+}
+write(i);
+write(j);
